src/test/tst_attrsgenerator.cpp

```diff
@@ -32,60 +32,321 @@ private slots:
+{
+   if (mode == "min") {
+        for (int i = 0; i < res.size(); ++i) {
+            for (int j = 0; j < numOfAttrRges; ++j) {
+                QCOMPARE(res.at(i).value(j), ascope.value(res.at(i).name(j))->min());
+            }
+        }
+    } else if (mode == "max") {
+        for (int i = 0; i < res.size(); ++i) {
+            for (int j = 0; j < numOfAttrRges; ++j) {
+                QCOMPARE(res.at(i).value(j), ascope.value(res.at(i).name(j))->max());
+            }
+        }
+    } else if (mode == "rand") {
+        for (int i = 0; i < res.size(); ++i) {
+            for (int j = 0; j < numOfAttrRges; ++j) {
+                QVERIFY(res.at(i).value(j) >= ascope.value(res.at(i).name(j))->min());
+                QVERIFY(res.at(i).value(j) <= ascope.value(res.at(i).name(j))->max());
+            }
+        }
+    }
+}
+
+AttributesScope TestAttrsGenerator::_newAttrsScope(const QStringList& names, const QStringList& attrRges)
+{
+    Q_ASSERT(names.size() == attrRges.size());
+    AttributesScope attrsScope;
+    attrsScope.reserve(names.size());
+    for (int i = 0; i  < names.size(); ++i) {
+        AttributeRange* a = AttributeRange::parse(i, names.at(i), attrRges.at(i));
+        attrsScope.insert(a->attrName(), a);
+    }
+    return attrsScope;
+}
+
+void TestAttrsGenerator::_tst_parseStarCmd(QString mode)
+{
+    AttributesScope attrsScope;
+    AttrsGenerator* agen;
+    SetOfAttributes res;
+    Attributes attrs;
+    const int sizeOfAgen = 3;
+    const QStringList names = {"test0", "test1", "test2"};
+    const QStringList attrRgeStrs = {"int[0,2]", "double[2.3,7.8]", "int{-2,0,2,4,6}"};
+    const QString seed("_123");
+    QString error, cmd;
+
+    // Valid * command with empty attrScope
+    if (!strcmp(Value(mode).toString(), "rand")) {
+        cmd = QString("*%1;%2%3").arg(Value(sizeOfAgen).toQString()).arg(mode).arg(seed);
+    } else {
+        cmd = QString("*%1;%2").arg(Value(sizeOfAgen).toQString()).arg(mode);
+    }
+
+    agen = AttrsGenerator::parse(attrsScope, cmd, error);
+
+    QCOMPARE(agen->command(), cmd);
+    QCOMPARE(agen->size(), sizeOfAgen);
+
+    res = agen->create();
+
+    _tst_attrs(res, attrs, true);
+
+    // Valid * command with non-empty attrScope
+    attrsScope = _newAttrsScope(names, attrRgeStrs);
+    agen = AttrsGenerator::parse(attrsScope, cmd, error);
+
+    attrs.resize(3);
+    for (int i = 0; i < attrs.size(); ++i) {
+        attrs.replace(i, names.at(i), NULL);
+    }
+
+    QCOMPARE(agen->command(), cmd);
+    QCOMPARE(agen->size(), sizeOfAgen);
+
+    res = agen->create();
+
+    _tst_attrs(res, attrs, false);
+    _tst_mode(res, mode, attrsScope, 3);
+}
+
 void TestAttrsGenerator::tst_parseStarCmd()
 {
+    _tst_parseStarCmd("min");
+    _tst_parseStarCmd("max");
+    _tst_parseStarCmd("rand");
+}
+
+void TestAttrsGenerator::tst_parseHashCmd_min()
```

Comment:
See https://github.com/evoplex/evoplex/pull/16#discussion_r205514555
As you did for the "star command", it seems that your `min` `max` and `rand` tests for the "hash command" could be simplified...